libs/langchain-community/src/vectorstores/turbopuffer.ts

```diff
@@ -0,0 +1,280 @@
+    return response;
+  }
+
+  async addVectors(
+    vectors: number[][],
+    documents: DocumentInterface[],
+    options?: { ids?: number[] }
+  ): Promise<void> {
+    try {
+      if (options?.ids && options.ids.length !== vectors.length) {
+        throw new Error(
+          "Number of ids provided does not match number of vectors"
+        );
+      }
+
+      if (documents.length !== vectors.length) {
+        throw new Error(
+          "Number of documents provided does not match number of vectors"
+        );
+      }
+
+      if (documents.length === 0) {
+        throw new Error("No documents provided");
+      }
+
+      const batchedVectors = chunkArray(vectors, this.batchSize);
+      const batchedDocuments = chunkArray(documents, this.batchSize);
+      const batchedIds = options?.ids
+        ? chunkArray(options.ids, this.batchSize)
+        : batchedDocuments.map((docs, index) =>
+            docs.map((_, docIndex) => index * this.batchSize + docIndex)
+          );
+
+      const batchRequests = batchedVectors.map(async (batchVectors, index) => {
+        const batchDocs = batchedDocuments[index];
+        const batchIds = batchedIds[index];
+
+        if (batchIds.length !== batchVectors.length) {
+          throw new Error(
+            "Number of ids provided does not match number of vectors"
+          );
+        }
+
+        const attributes = {
+          pageContent: batchDocs.map((doc) => doc.pageContent),
+          metadata: batchDocs.map((doc) => JSON.stringify(doc.metadata ?? {})),
+        };
+
+        const data = {
+          ids: batchIds,
+          vectors: batchVectors,
+          attributes,
+        };
+
+        const response = await this.callWithRetry(
+          `${this.apiUrl}/vectors/${this.namespace}`,
+          JSON.stringify(data)
+        );
+
+        if (response.status !== 200) {
+          throw new Error(
+            `Failed to add vectors to Turbopuffer. Response status ${
+              response.status
+            }\nFull response: ${await response.text()}`
+          );
+        }
+      });
+
+      // Execute all batch requests in parallel
+      await Promise.all(batchRequests);
+    } catch (error) {
+      console.error("Error storing vectors:", error);
+      throw error;
+    }
+  }
+
+  async addDocuments(
+    documents: DocumentInterface[],
+    options?: { ids?: number[] }
+  ): Promise<void> {
+    const vectors = await this.embeddings.embedDocuments(
+      documents.map((doc) => doc.pageContent)
+    );
+
+    return this.addVectors(vectors, documents, options);
+  }
+
+  async queryVectors(
+    query: number[],
+    k: number,
+    includeAttributes?: string[],
+    includeVector?: boolean,
+    // See https://Turbopuffer.com/docs/reference/query for more info
+    filter?: TurbopufferFilterType
+  ): Promise<TurbopufferQueryResult[]> {
+    const data = {
+      vector: query,
+      top_k: k,
+      distance_metric: this.distanceMetric,
+      filters: filter,
```

Comment:
We store document contents in the attributes field above as:
```
        const attributes = {
          pageContent: batchDocs.map((doc) => doc.pageContent),
          metadata: batchDocs.map((doc) => JSON.stringify(doc.metadata ?? {})),
        };
```

Could be why passing `filter` through directly isn't working? We could also store raw document content as something like `__lc_page_content` and make sure we handle it while querying and inserting.